File: LinkListCycle.go
package main

func hasCycle(head *ListNode) bool {
	if head == nil || head.Next == nil {
		return false
	}
	fast := head.Next
	slow := head

	for fast != slow {
		if fast == nil || fast.Next == nil {
			return false
		} else {
			fast = fast.Next.Next
			slow = slow.Next
		}
	}
	return true

}

File: LinkListCycle.py

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        if not (head and head.next):
            return False 
        
        # 快慢指针, 环形链表终有一个点快慢指针可以交汇
        fast, slow = head.next, head
        while fast != slow:
            if not (fast and fast.next):
                # 链表到达了终点
                return False 
            else:
                fast = fast.next.next 
                slow = slow.next        
        return True # 交汇后返回True

        



